[IV-24-25] Objetivo3

abbonno opened the pull request:
* [x] ¿Incluye el fichero `iv.yaml` las claves de todos los objetivos
anteriores?
automatizar:
  fichero: Taskfile.yml
  orden: task
* [x] ¿Se ha desarrollado el fichero para las tareas a partir de un
   issue?
* [x] ¿Ese issue está asignado a un milestone?
Sí, se asigna al milestone 0.
* [x] ¿El milestone incluye, entre sus productos mínimamente viables, la
  existencia de herramientas que permitan automatizar tareas?
Sí, el milestone pide decidir las tecnologías a emplear para llevar a cabo el desarrollo del proyecto.
* [x] ¿Hemos justificado correctamente en caso de elegir un gestor de
   dependencias el que se use como gestor de tareas?
* [x] ¿El PR, que también es un issue, está asignado al mismo milestone?
* [x] ¿La tarea `check` efectivamente comprueba la sintaxis de los fuentes
  existentes y sólo eso?
* [x] ¿Se ha comprobado que las instrucciones incluidas en el `README.md` son
completas y correctas?

abbonno committed:
Preparando pr Objetivo3

abbonno committed:
Crea documentación sobre el gestor de dependencias #17

abbonno committed:
Crea documentación sobre elección del gestor de tareas #16

abbonno committed:
Añade la información necesaria a iv.yml #18

abbonno committed:
Crea el fichero de automatización de tareas y añade información respectiva al readme

abbonno committed:
Elimina función innecesaria de solución de problemas, check solo debe de comprobar la sintaxis #19

abbonno commented:
@JJ revisión

JJ commented on docs/gestor_dependecias.md:
```diff
@@ -0,0 +1,7 @@
+
+## SELECCIÓN FINAL
+
+    **GO MOD**
```
No va en mayúsculas, pero bueno.

abbonno replied:
[ce7d8cf]

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,58 @@
+
+    Se emplearán los siguientes criterios en la selección de una de las herramientas exploradas para gestionar las tareas del proyecto.
+
+    - 1. Se valorarán positivamente las herramientas que no añadan complejidad al proyecto (es preferible usar una específica, evitando aprender nuevos lenguajes, pero si es más complicado que escribir un comando sencillo en otro lenguaje se elegirá uno genérico).
```
¿Es objetiva la complejidad? ¿Es medible? Si lo es, ¿por qué no pones un criterio específico que sí lo sea?

abbonno replied:
Es objetivo ya que no tiene en cuenta el conocimiento previo del desarrollador. En primer lugar tienen preferencia las herramientas escritas en go (siguiendo la línea del proyecto), en segundo lugar las que usan código más cercano al lenguaje natural (python, yaml...) y por último el resto.

JJ replied:
Ni lo que comentas tiene nada que ver con la complejidad, ni es objetivo. Y si es eso lo que quieres decir, dilo. Que permita escribir las tareas en Go es objetivo (o lo permite o no). Que añada o reste complejidad, es discutible. ¿Cómo haces `rm -rf` en Go?

abbonno replied:
Sí es cierto que la complejidad puede ser algo ambiguo y, consecuentemente, subjetivo, pero en este caso puede ser medible. El ejemplo que me has puesto es adecuado para la condición especificada "si es más complicado que escribir un comando sencillo en otro lenguaje se elegirá uno genérico". Para hacer "rm -rf" harían falta más líneas de código y sería menos legible, algo que lo hace objetivamente peor aún conociendo el lenguaje.

abbonno replied:
[1806d69]

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,58 @@
+
+    - 1. Se valorarán positivamente las herramientas que no añadan complejidad al proyecto (es preferible usar una específica, evitando aprender nuevos lenguajes, pero si es más complicado que escribir un comando sencillo en otro lenguaje se elegirá uno genérico).
+    - 2. Actualizado: tendrán preferencia los gestores de dependencias que no hayan sido "abandonados" o no reciban modificaciones de forma habitual, por lo que tienen un buen mantenimiento. De la misma manera, aunque no sean actualizados continuamente serán bien consideradas las herramientas que se encuentren en etapas de desarrollo muy estables y, por lo tanto, sin errores críticos.
+    - 3. Herramienta concisa: sirve para lo que sirve, evitando lo especificado en los errores comunes *"Usar herramientas que hacen más cosas, incluso mucho más cosas, que verificar la sintaxis. Se trata simplemente de ver si el código compila o no compila, no si sigue las mejores prácticas y demás."*
```
Esa frase no se refiere al gestor de tareas, sino al comando que comprueba la sintaxis.

abbonno replied:
Entiendo. Decido no tenerlo en cuenta. [92c7645]

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,58 @@
+    - 1. Se valorarán positivamente las herramientas que no añadan complejidad al proyecto (es preferible usar una específica, evitando aprender nuevos lenguajes, pero si es más complicado que escribir un comando sencillo en otro lenguaje se elegirá uno genérico).
+    - 2. Actualizado: tendrán preferencia los gestores de dependencias que no hayan sido "abandonados" o no reciban modificaciones de forma habitual, por lo que tienen un buen mantenimiento. De la misma manera, aunque no sean actualizados continuamente serán bien consideradas las herramientas que se encuentren en etapas de desarrollo muy estables y, por lo tanto, sin errores críticos.
+    - 3. Herramienta concisa: sirve para lo que sirve, evitando lo especificado en los errores comunes *"Usar herramientas que hacen más cosas, incluso mucho más cosas, que verificar la sintaxis. Se trata simplemente de ver si el código compila o no compila, no si sigue las mejores prácticas y demás."*
+    - 4. Peso: si la herramienta necesita muchas dependencias externas u ocupa demasiado espacio en su instalación se verá reflejado negativamente. Además, es probable que, si añade mucho contenido no cumpla el criterio anterior.
```
¿Esto está relacionado con la complejidad o no?

abbonno replied:
No necesariamente, si una herramienta requiere de generar mucho código para tareas sencillas es compleja y además añade archivos más pesados, pero en mi criterio hablo únicamente del caso de la instalación y sus dependencias.

abbonno replied:
[1806d69]

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,58 @@
+
+## OPCIONES
+
+### Go build
```
¿Esto gestiona tareas?

abbonno replied:
No, se pone a título informativo sobre por qué no se usa aún siendo una build tool oficial de go.

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,58 @@
+
+### Go build
+
+    Puede ser sencillo confundir en los foros y fuentes de información *task managers* (gestor de tareas) y *build tools* (herramientas de construcción). Esto se debe a que existen muchas herramientas de construcción que llegan a cumplir tareas de testing, generación de documentación y más, a parte llevar a cabo su compilación. Pero muchas otras solo cumplen esta última, por lo que no nos sirven. ...
```
¿O menos por no ser un gestor de tareas?

abbonno replied:
Efectivamente, es por esa razón que se descarta o más bien, ni siquiera se plantea su uso.

JJ reviewed (changes requested):
Por favor, trata de entender el concepto de gestor de tareas antes de listar herramientas que no lo sean. Que build tools como `make` eventualmente evolucionen hacia ser un gestor de tareas no implica que *todos* lo hagan.

abbonno committed:
Corrige go mod en minúsculas #17

abbonno committed:
Corrige criterios selección gestor de tareas #16

abbonno committed:
Suprime complejidad y peso como criterios #16

La complejidad es un criterio no objetivo y el peso no es importante, así que no cabe tenerlos en cuenta como criterios de selección

abbonno commented:
@JJ revisión

JJ reviewed (changes requested):
Como pone en [los errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/objetivo-3.md),

> n este artículo, como en todos, se trata de que busquéis y trabajéis, no que copiéis lo que han hecho otros, lo que os diga la IA, o lo que salga en la primera página de Google. Si sólo "encuentras" make, mage y task, busca un poco más, porque hay muchas más herramientas.

abbonno committed:
Añade opciones para comparar gestores de tareas #16

Se añaden a la comparativa buck2, rake y ninja. Se mantiene la elección de Task

abbonno commented:
@JJ revisión.
Ya he añadido varias comparativas (ese error frecuente no estaba antes de realizar mi último commit, me pondré la medallita de haber contribuido).

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,51 @@
+
+    Se emplearán los siguientes criterios en la selección de una de las herramientas exploradas para gestionar las tareas del proyecto.
+
+    - Actualizado: tendrán preferencia los gestores de dependencias que no hayan sido "abandonados" o no reciban modificaciones de forma habitual, por lo que tienen un buen mantenimiento. De la misma manera, aunque no sean actualizados continuamente serán bien consideradas las herramientas que se encuentren en etapas de desarrollo muy estables y, por lo tanto, sin errores críticos.
```
Esto ¿son dos criterios diferentes o sólo uno?

abbonno replied:
Podría separarlo en dos perfectamente, gracias por la anotación.

abbonno replied:
[ade091e]

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,51 @@
+### Rake
+
+    Como otras herramientas a las que solo se cambia una letra a partir de "Make" (mage, rake, cake...), es un Make-like pero implementado en Ruby.
+    Su comunidad no es muy activa en comparación con las herramientas anteriores y esto sería justificable por el hecho de no tener que lidiar con muchos issues debido a su desarrollo anterior, habiendo comenzado en 2008.
```
No es muy activa y además, ¿necesita Ruby? Si vas a meter herramientas que necesiten otros lenguajes hay un buen puñado...

abbonno replied:
Como finalmente no consideramos el que esté en otro lenguaje como condicionante no supone un impedimento para reflexionar su uso. Lo he tenido en cuenta como podía haber tenido los otros muchos que existen.

JJ replied:
El problema es que si te abres a otros lenguajes, tienes que abrirte a *todos*. Y meter uno *de otro lenguaje* y que además no es muy activo no sé si te merece la pena. En todo caso, tienes que resolver los comentarios *siempre* con un cambio. Eso deberías saberlo ya.

abbonno replied:
Entendido, he desechado Rake como opción. Si ves que es necesario añadir alguna comparativa más, tengo varias, pero creo que con las que hay es suficiente. [5408d29]

JJ reviewed (changes requested):
No metas "todo lo que tiene que tener para que me guste" en un solo criterio. Si son criterios diferentes, son diferentes.
Pon también un enlace a las herramientas que menciones.

abbonno committed:
Arregla criterios de selección del gestor de tareas

El criterio existente se puede dividir en dos independientes: actualidad/mantenimiento y estabilidad/madurez de la herramienta

abbonno commented:
@JJ revisión

abbonno commented:
> No metas "todo lo que tiene que tener para que me guste" en un solo criterio. Si son criterios diferentes, son diferentes. Pon también un enlace a las herramientas que menciones.

Acabo de recordar lo de los enlaces, no lo acabé la primera vez.

abbonno committed:
Añade enlaces a build tools (sus páginas oficiales) #16

abbonno committed:
Arregla formato markdown gestor tareas y gestor dependencias

abbonno committed:
Descarta gestores de tareas en otros lenguajes

Se descarga rake, que está en ruby

abbonno committed:
Descarta gestores de tareas en otros lenguajes (correccion) #16

abbonno commented:
@JJ

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,47 @@
+
+### [Task](https://taskfile.dev/)
+
+Desde el mismo foro se habla de Task. Este, igualmente, según su documentación, dice diferenciarse de make por su simplicidad.
```
¿Qué foro?

abbonno replied:
Reddit

JJ replied:
¿Qué commit?

abbonno replied:
[fc386d4] Se modifica la mención al foro. Daba ciertas ideas sobre herramientas, pero el sujeto del mismo eran build tools y no herramientas de gestión de proyectos: https://www.reddit.com/r/golang/comments/19dhlx1/what_build_tools_do_you_use/?rdt=34155

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,47 @@
+
+### [Buck2](https://buck2.build/)
+
+Buck2 es la build tool desarrollada y usada por Meta. Está escrita en Rust y en su página principal remarcan el uso de ejecución remota y construcción incremental guardando en caché las instrucciones anteriormente compiladas.
```
¿Una build tool es un gestor de tareas?

abbonno replied:
No, no lo es. Aunque esta fuera su descripción pensé que podría considerarla, ya que aún así podría servir para automatizar tareas, pero explorando más a fondo veo que no. La suprimiré junto a Ninja, por las mismas razones.

abbonno replied:
[e4ec79e]

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,47 @@
+Es una herramientam muy reciente, con solo dos años de desarrollo, aunque parte de buck1 que comenzó a ser desarrollada en 2014. Su comunidad es activa y recibe actualizaciones a diario.
+La propia página oficial admite como desventaja su estabilidad: "Buck2 está en desarrollo activo, lo que significa que el riesgo de regresión es correspondientemente mayor. Puede haber problemas, pero se solucionarán lo más rápido posible (y se aprenderán lecciones) a través del proceso de revisión SEV de Meta."
+
+### [Ninja](https://ninja-build.org/)
```
¿Esto es un gestor de tareas? ¿Cómo se define en su web?

abbonno replied:
Ídem.

abbonno replied:
[e4ec79e]

JJ reviewed (changes requested):
Y una vez que descartes los que (posiblemente) no correspondan al tipo de herramienta que estás buscando, mira los [errores comunes](https://github.com/JJ/IV-/blob/master/errores/objetivo-3.md):

> En este objetivo, como en todos, se trata de que busquéis y trabajéis, no que copiéis lo que han hecho otros, lo que os diga la IA, o lo que salga en la primera página de Google. Si sólo "encuentras" make, mage y task, busca un poco más, porque hay muchas más herramientas.

abbonno committed:
Actualización gestores de tareas debatidos #16

Se eliminan las herramientas ninja y buck2 al no presentar las características de gestor de tareas y se añaden dos que sí las cumplen.

abbonno commented:
@JJ

JJ reviewed (commented):
Por favor, dale una pequeña vuelta.

abbonno committed:
Descarta ansible #16

Descartamos ansible, no es un gestor de tareas, añadimos Just, que sí lo es, añadiendo mejoras a make y actualizado

abbonno commented:
@JJ

abbonno committed:
Añade enlaces a github como referencia #16

JJ commented on Taskfile.yml:
```diff
@@ -0,0 +1,7 @@
+  check:
+    desc: Validar la sintaxis de los archivos fuente de Go.
+    cmds:
+      - go vet ./...
```
¿Qué hace go vet?

abbonno replied:
"vet examina el código fuente de Go y reporta construcciones sospechosas, así como llamadas Printf cuyos argumentos no se alinean con el formato del string. Vet usa heurística que no garantiza que todos los informes sean problemas genuinos, pero puede encontrar errores no encontrados por los compiladores". De modo que no realiza la función que queremos. Cambio la función por la siguiente: gofmt y su función "-e", que muestra los errores de sintáxis.

PS C:\Users\AntonioGonzález\Desktop\info\IV\healthScheduler> gofmt --help
usage: gofmt [flags] [path ...]
  -cpuprofile string
        write cpu profile to this file
  -d    display diffs instead of rewriting files
  **-e    report all errors (not just the first 10 on different lines)**
  -l    list files whose formatting differs from gofmt's
  -r string
        rewrite rule (e.g., 'a[b:len(a)] -> a[b:]')
  -s    simplify code
  -w    write result to (source) file instead of stdout

[311ff37]

JJ commented on docs/gestor_tareas.md:
```diff
@@ -0,0 +1,48 @@
+### [Goyek](https://github.com/goyek/goyek)
+
+Se trata de una librería de automatización de tareas en go que permite organizar flujos de trabajo mediante funciones de este lenguaje.
+Como proyecto es bastante simple y con muchos menos colaboradores, por ende, con menos mantenimiento. Además tiene su origen en 2021.
```
¿"Además" es bueno o malo? ¿Es bueno que se hiciera después de la pandemia?

abbonno replied:
Arreglé la redacción [1e33c06]

abbonno replied:
Olvidé quitar una línea referente a buck2 ya eliminado. Corrijo aquí [a1a7dcf]

JJ reviewed (commented):
por favor, que compruebe *solamente* la sintaxis.

abbonno committed:
Configura correctamente la función check #19

Se ha explorado la función de gofmt y es correcta para este caso

abbonno committed:
Mejora redacción documentación gestor de tareas #16

abbonno committed:
Mejora redacción selección final #16

Aún quedaban residuos de una antigua elección

abbonno commented:
@JJ revisión

JJ reviewed (approved):
:+1:
